Manylinux2014_Compliant_Source/pkg/IR2Vec/core.cpp

```diff
@@ -47,342 +51,366 @@
+  }
 
-    if (temp2.size() != 0) {
-      temp2.clear();
+  // generateEncodings
+  PyObject *generateEncodings(OpType type, std::string funcName = "") {
+    // Invokinng IR2Vec lib exposed functions
+    IR2Vec::iname = this->fileName;
+    IR2Vec::IR2VecMode ir2vecMode =
+        (this->mode == string("sym") ? IR2Vec::Symbolic : IR2Vec::FlowAware);
+    // The scope of this Module object is extremely crucial
+    std::unique_ptr<llvm::Module> Module;
+    Module = IR2Vec::getLLVMIR();
+    std::string vocab_path = seed_emb_path + "/seedEmbeddingVocab-llvm14.txt";
+
+    IR2Vec::Embeddings *emb = new IR2Vec::Embeddings();
+    // if output file is provided
+    if (this->outputFile != "") {
+      string outFile = this->outputFile;
+      ofstream output;
+      output.open(outFile, ios_base::app);
+      emb = std::move(new IR2Vec::Embeddings(*Module, ir2vecMode, vocab_path,
+                                             (this->level)[0], &output,
+                                             funcName));
+    } else {
+      emb = std::move(new IR2Vec::Embeddings(*Module, ir2vecMode, vocab_path,
+                                             (this->level)[0], nullptr,
+                                             funcName));
     }
 
-    // for FuncVecMap
-    for (auto Func_it : llvm_func_vec_map) {
-      // copy this SmallVector into c++ Vector
-      for (auto &Vec_it : Func_it.second)
-        temp2.push_back(Vec_it);
-
-      // apply __cxx::demangle just to be cautious
-      demangledName = IR2Vec::getDemagledName(Func_it.first);
-      FuncVecMap[demangledName] = temp2;
+    if (emb == nullptr) {
+      PyErr_SetString(PyExc_TypeError, "Embedding Object not created");
+      Py_RETURN_NONE;
+    }
 
-      temp2.clear();
+    if (type == OpType::Program) {
+      llvm::SmallVector<double, DIM> progVector = emb->getProgramVector();
+      return this->createProgramVectorList(progVector);
+    } else if (type == OpType::Function) {
+      llvm::SmallMapVector<const llvm::Function *, IR2Vec::Vector, 16>
+          funcVecMap = emb->getFunctionVecMap();
+      return this->createFunctionVectorDict(funcVecMap);
+    } else if (type == OpType::Instruction) {
+      llvm::SmallMapVector<const llvm::Instruction *, IR2Vec::Vector, 128>
+          instVecMap = emb->getInstVecMap();
+      return this->createInstructionVectorDict(instVecMap);
+    } else {
+      PyErr_SetString(PyExc_TypeError, "Invalid OpType");
+      Py_RETURN_NONE;
     }
+  }
+};
 
-    // for PgmVector
-    for (auto &Pgm_it : llvm_pgm_vec)
-      PgmVect.push_back(Pgm_it);
+typedef struct {
+  PyObject_VAR_HEAD ir2vecHandler *ir2vecObj;
+} ir2vecHandlerObject;
 
-    // all c++ container are ready to convert
-    // iterate over a c++ container , if it has embedd anotehr c++c ocntainer
-    // then iterate over it also
+PyObject *getInstructionVectors(ir2vecHandlerObject *self, PyObject *args) {
+  // check for args, and null etc
+  if ((self->ir2vecObj) == NULL) {
```

Comment:
Should these checks be on NULL or nullptr?

Reply:
Please check. These checks are in multiple places.

Reply:
My bad. 

I've pushed the following changes here
1. I've replaced `NULL` with `nullptr` wherever pointer comparisons are taking place, or where I expect deleted pointers.
2. Replacing instance of NULL with Py_RETURN_NONE wherever we're returning a None Type object to the Python API